### PR TITLE
Datetime support for 1-way Partial Dependence

### LINKED ISSUE #2385: Speed up partial dependence for date-time features
In PR #2365, we noticed that one of our slowest tests are those that compute partial dependence on datetimes. The reason is that those tests had a date feature with 100 unique dates, and our current partial dependence implementation will compute the partial dependence on every unique date. This means that the `grid_resolution` parameter is not respected for dates and will be slow for datasets with many unique dates.

The underlying problem is that scikit-learn's partial_dependence implementation computes the grid for you and it cannot compute a grid from date features. 

I think it's time we move to our own partial dependence implementation so that we can control how the grid is computed. I sketched out how this could work in the original partial-dependence-for-dates issue #1963 

https://github.com/alteryx/evalml/issues/1963#issuecomment-810628780

### PR DESCRIPTION
**Pull request by bchen1116**
fix #2385 

We test partial dependence with the fraud dataset (1000 rows only), using the Elastic Net Classifier estimator

**Testing the runtime for the new implementation:**
image
Output:
image

**Comparing the runtime on the main branch:**
image
image
image
We cut all of these tests short since it takes about 5-7 minutes per iteration here. 

Overall, we see we have a drastic improvement in runtime

### CODE REVIEW

File: evalml/tests/model_understanding_tests/test_partial_dependence.py [hunk 1]
```diff
@@ -965,28 +967,37 @@ def test_partial_dependence_datetime(
...
-        assert len(part_dep["partial_dependence"]) == 10
-        assert len(part_dep["feature_values"]) == 10
+        assert len(part_dep["partial_dependence"]) == expected_size
+        assert len(part_dep["feature_values"]) == expected_size
     assert not part_dep.isnull().any(axis=None)
-
-    part_dep = partial_dependence(pipeline, X, features=20)
+    # keeps the test from running too long. The part below still runs for 3 other tests
+    if grid == 100 or size == 100:
+        return
```
Comment (bchen1116) [inline evalml/tests/model_understanding_tests/test_partial_dependence.py:991]: This test takes about 1 min to run locally.

File: evalml/model_understanding/graphs.py [hunk 2]
```diff
@@ -690,15 +691,51 @@ def partial_dependence(
...
+        if any(is_datetime):
+            timestamps = np.array(
+                [X_dt - pd.Timestamp("1970-01-01")] // np.timedelta64(1, "s")
+            ).reshape(-1, 1)
+            grid, values = _grid_from_X(
+                timestamps, percentiles=percentiles, grid_resolution=grid_resolution
+            )
+            grid_dates = pd.to_datetime(
+                pd.Series(grid.squeeze()), unit="s"
+            ).values.reshape(-1, 1)
```
Comment (chukarsten) [inline evalml/model_understanding/graphs.py:705]: I like this, I think it's clever.  Converting to seconds after Unix time, yes?  So ultimately your using a large integer number to compute the grid.  Perhaps we can use this calculation to replace the dates with seconds after unix time, run the data through sklearn's stock partial dependence, then convert the datetime feature back after the fact?  To me, I think that would mitigate some of the code duplication and use of private funcs issues.

**Review by chukarsten — CHANGES_REQUESTED** [review]
Nice job overcoming the issue and getting the performance increase.  I think you're definitely on the right track, but I do have some concerns about using the sklearn private functions as well as duplicating some of the code in sklearn's partial_dependence().  I'd like to explore maybe modifying the data so that you can send it into the stock part_dep function and then perhaps fix it after it comes out.  Not sure if that's a good solution, though so would be happy to pair on it.

**Comment by bchen1116** [comment]
@chukarsten filed an issue to track fixing/cleaning this up [here](https://github.com/alteryx/evalml/issues/2502)

**Review by freddyaboulton — APPROVED** [review]
@bchen1116 Thanks for doing this! The speed-up looks fantastic. I left some suggestions for adding test coverage. Definitely agree that the next step is to clean up our partial dependence - possible roll our own implementation!

File: evalml/pipelines/components/transformers/preprocessing/datetime_featurizer.py [hunk 3]
```diff
@@ -133,6 +135,10 @@ def transform(self, X, y=None):
         for col_name in self._date_time_col_names:
             for feature in features_to_extract:
                 name = f"{col_name}_{feature}"
+                # if is_numeric_dtype(X[col_name].dtype):
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/pipelines/components/transformers/preprocessing/datetime_featurizer.py:138]: Good catch!

File: evalml/tests/model_understanding_tests/test_partial_dependence.py (hunk 1 above)
Comment (bchen1116) [reply evalml/tests/model_understanding_tests/test_partial_dependence.py:992]: @freddyaboulton The entire test takes about 1 minute locally! Leaving the grid and size unconstrained makes it run much longer

File: evalml/model_understanding/graphs.py [hunk 4]
```diff
@@ -690,15 +691,51 @@ def partial_dependence(
...
-            wrapped,
-            X=X,
-            features=features,
-            percentiles=percentiles,
-            grid_resolution=grid_resolution,
-            kind=kind,
-        )
+        if any(is_datetime):
+            timestamps = np.array(
+                [X_dt - pd.Timestamp("1970-01-01")] // np.timedelta64(1, "s")
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/model_understanding/graphs.py:698]: @freddyaboulton updated the tests!

[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/model_understanding/graphs.py:698]: Thanks @bchen1116 Looks good to me! Is the runtime reasonable?

[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/model_understanding/graphs.py:698]: @freddyaboulton It takes less than 20 seconds locally to run

**Pull request merged**